src/Lib/BSP/IO/LumpReader.cs

```diff
@@ -0,0 +1,101 @@
+        public LumpReader(Stream input) : base(input)
+        { }
+        protected abstract void ReadHeader();
+        protected virtual void LoadAll()
+        {
+            foreach (var l in Lumps)
+            {
+                var lump = l.Item1;
+                var lumpHeader = l.Item2;
+                if (lumpHeader.Length > 0)
+                    Read(lump, lumpHeader);
+            }
+        }
+        private MemoryStream Decompress()
+        {
+            MemoryStream decompressedStream = new();
+
+            const string magic = "LZMA";
+            var id = ReadBytes(magic.Length);
+            if (Encoding.ASCII.GetString(id) != magic)
+                throw new InvalidDataException("Failed to decompress: Lump doesn't look like it was LZMA compressed");
+
+            uint actualSize = ReadUInt32();
+            uint lzmaSize = ReadUInt32();
+            byte[] properties = new byte[5];
+
+            Read(properties, 0, 5);
+
+            var lzmaStream = new LzmaStream(properties,
+                                            BaseStream,
+                                            lzmaSize,
+                                            actualSize);
+            lzmaStream.CopyTo(decompressedStream);
+            decompressedStream.Flush();
+            decompressedStream.Seek(0, SeekOrigin.Begin);
+
+            return decompressedStream;
+        }
+        protected void Read(Lump lump, LumpHeader lumpHeader)
+        {
+            var decompress = lump is not IUnmanagedLump;
+
+            BinaryReader lumpReader;
+            long lumpStreamLength;
+
+            BaseStream.Seek(lumpHeader.Offset, SeekOrigin.Begin);
+
+            if (lumpHeader.Compressed)
+            {
+                if (decompress)
+                {
+                    MemoryStream decompressedStream = Decompress();
+                    lumpReader = new BinaryReader(decompressedStream);
+                    lumpStreamLength = decompressedStream.Length;
+                    if (lump is IUnmanagedLump unmanagedLump)
+                    {
+                        unmanagedLump.Compressed = false;
+                        unmanagedLump.UncompressedLength = -1;
+                    }
+                }
+                else if (lump is IUnmanagedLump unmanagedLump)
+                {
+                    unmanagedLump.Compressed = true;
+                    unmanagedLump.UncompressedLength = lumpHeader.UncompressedLength;
+                    lumpReader = this;
+                    lumpStreamLength = lumpHeader.CompressedLength;
+                }
+                else
+                    throw new NotImplementedException(
+                        "not decompressing a managed lump not implemented .."
+                        + " make it unmanaged if you don't care about it or decompress and read");
+            }
+            else
+            {
+                if (lump is IUnmanagedLump unmanagedLump)
```

Comment:
Looks like we should reuse the `decompress` bool here. Though I think it would be clear to name that variable something like `shouldDecompress`.

Reply:
I need the unmanagedLump variable here so I thought I just make it like this instead of casting and if we would change how decompress gets its value this would break.
Why is decompress not clear?. I don't mind changing it but then I should also change "Compress" in the lump class to "ShouldCompress" and "Compressed" to "IsCompressed" so its consistent, right?

Reply:
Mmh okay. My issue is just that doing this `is IUnmanagedLump unmanagedLump` multiple times, storing the bool in one case but then doing the same check again without using the stored bool, it makes the overall logic a little confusing. I personally like using language that makes it clear where things are bools but that's really just preference I guess.

In this case I think just removing the `decompress` bool is best, it's only used once. But maybe I'm being overly nitpicky about it, once the unreachable code is gone it should read much better anyway